wasm.md

```diff
@@ -596,36 +599,86 @@ The `*_local` instructions are defined here.
     rule <k> ( global.set INDEX INSTR ) => INSTR ~> ( global.set INDEX ) ... </k>
 ```
 
-Function Declaration and Invocation
------------------------------------
+Types
+-----
 
-### Function Types Gathering
+### Type Gathering
 
-```k
-    syntax TypeKeyWord ::= "param" | "result" | "local"
- // ---------------------------------------------------
+This defines helper functions that gathers function together.
 
-    syntax FuncDecl  ::= "(" FuncDecl ")"     [bracket]
-                       | TypeKeyWord ValTypes
-    syntax FuncDecls ::= List{FuncDecl, ""} [klabel(listFuncDecl)]
- // --------------------------------------------------------------
+```k
+    syntax TypeKeyWord  ::= "param" | "result"
+ // ------------------------------------------
 
-    syntax FuncType ::= gatherFuncType ( FuncDecls ) [function]
- // -----------------------------------------------------------
-    rule gatherFuncType(FDECLS) => gatherTypes(param, FDECLS) -> gatherTypes(result, FDECLS)
+    syntax TypeDecl      ::= "(" TypeDecl ")"     [bracket]
+                           | TypeKeyWord ValTypes
+    syntax TypeDecls     ::= List{TypeDecl , ""} [klabel(listTypeDecl)]
+ // -------------------------------------------------------------------
 
-    syntax VecType ::=  gatherTypes ( TypeKeyWord , FuncDecls )            [function]
-                     | #gatherTypes ( TypeKeyWord , FuncDecls , ValTypes ) [function]
+    syntax VecType ::=  gatherTypes ( TypeKeyWord , TypeDecls )            [function]
+                     | #gatherTypes ( TypeKeyWord , TypeDecls , ValTypes ) [function]
  // ---------------------------------------------------------------------------------
-    rule gatherTypes(TKW, FDECLS) => #gatherTypes(TKW, FDECLS, .ValTypes)
+    rule  gatherTypes(TKW , TDECLS:TypeDecls) => #gatherTypes(TKW, TDECLS, .ValTypes)
+
+    rule #gatherTypes(TKW , .TypeDecls            , TYPES) => [ TYPES ]
+    rule #gatherTypes(TKW , TDECL:TypeDecl TDECLS , TYPES) => #gatherTypes(TKW, TDECLS, TYPES) [owise]
```

Comment:
Probably better to use `TKW' _` instead of `TDECL` and `requires TKW' =/=K TKW` than `[owise]`, if I understand correctly. Dwight and Everett have warned me against `[owise]`.